Breaks down the piano roll's toolbar into multiple smaller ones (#2286)

michaelgregorius opened the pull request:
Solves #2286 by grouping the piano roll's actions into several toolbars. There
are now five groups / tool bars: "Transport controls", "Note controls",
"Copy paste controls", "Timeline controls" and "Zoom and note controls"
(this group is a bit mixed). Each group can be turned off and on using
the standard toolbar context menu provided by Qt.

The new default layout of the toolbars saves horizontal space by putting
the "Zoom and note controls" below the other toolbars. The toolbars can
be moved but their positions are not saved. Floating toolbars are
disabled, i.e. the toolbars will always stay inside the window (top,
bottom, left, right).

![pianorollchanges-movabletoolsbars](https://cloud.githubusercontent.com/assets/9293269/9428756/25249618-49b9-11e5-9f12-4c9638e28cf1.png)

michaelgregorius commented:
I have switched the other editor to use movable toolbars as well. Here is what they look like:
![editor-toolbars](https://cloud.githubusercontent.com/assets/9293269/9450950/14b12442-4aac-11e5-99e4-2985d2826ee6.png)

tresf commented:
I'm a pretty big fan of this change. :+1:

grejppi commented:
Oh, great. Go a step further: you should add that hideous handle to every single button too. That way the user has the option to set it however they like.

No but seriously:
* What's the point of being able to move all of the toolbar?
* Does it enhance your productivity?
* Does it help your muscle memory?
* What's the point anyway if your precious ~customized~ workflow gets reset every time?

Now look at the second screenshot you posted. In the piano roll and automation editors the tool buttons and view controls sit nicely on their own lines, and the editors are hardly wider than the rest.

Wouldn't that be enough?

Wallacoloo commented:
Although a bit harsh, @grejppi does have a point - the toolbar handles are ugly and the config being reset on each start is sub-optimal.

One thing I think @grejppi may have missed is that this allows users to hide tools they don't use frequently (e.g. zoom/quantize), and this makes the interface cleaner and more compact for those users.

But I think perhaps it may be worth figuring out how to hide the handles before merging this PR. In the right-click menu you show, perhaps there could be an additional entry, "locked", which, when checked, removes the handles so everything looks clean. If you can do that in this PR, and we default "locked" to being checked for the time being, then this features provides a net benefit with essentially zero drawbacks. Config saving could then be dealt with in a future PR.

I'm not sure if the implementation method I suggested is possible within Qt. If not, we could also go the route of either:
* Handles are hidden by default. Long-pressing one of the buttons unhides its handle & allows it to be moved.
* Handles are hidden by default but buttons can still be rearranged by dragging them. Worst-case, this could be accomplished with just a style hack.

Personally, I prefer the context-menu "locked" entry as a solution.

tresf commented:
![image](https://cloud.githubusercontent.com/assets/6345473/9456007/fb933858-4a9f-11e5-84a2-20aa570c2445.png)

`QToolBar.setMovable(false);` would hide the drag handles, which I agree have no benefit.

My original attraction to the toolbar grouping was I thought it fixed the overflowing toolbar problem when the MDI windows were smaller than the toolbar.  After testing out the PR, I realize behavior is a bit different... when the toolbar is too small, it collapses onto itself, which is confusing with so many of them.

![image](https://cloud.githubusercontent.com/assets/6345473/9456031/5c4a2634-4aa0-11e5-80ca-d6d699137658.png)

I do like the ability to hide the sections, but I'm not sure I like the collapsing behavior.

> Oh, great.

You got your point across in the first section of your rant.  We're trying to solve problems here. :+1:

Spekular commented:
I think the ugly handles are a result of them being fully opaque. They're
meant to be semitransparent so they blend with the background like the
piano roll track handles.
On Aug 25, 2015 2:42 AM, "Tres Finocchiaro" <notifications@github.com>
wrote:

> [image: image]
> <https://cloud.githubusercontent.com/assets/6345473/9456007/fb933858-4a9f-11e5-84a2-20aa570c2445.png>
>
> QToolBar.setMovable(false); would hide the drag handles, which I agree
> have no benefit.
>
> My original attraction to the toolbar grouping was I thought it fixed the
> overflowing toolbar problem when the MDI windows were smaller than the
> toolbar. After testing out the PR, I realize behavior is a bit different...
> when the toolbar is too small, it collapses onto itself, which is confusing
> with so many of them.
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/6345473/9456031/5c4a2634-4aa0-11e5-80ca-d6d699137658.png>
>
> I do like the ability to hide the sections, but I'm not sure I like the
> collapsing behavior.
>
> Oh, great.
>
> You got your point across in the first section of your rant. We're trying
> to solve problems here. [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LMMS/lmms/pull/2287#issuecomment-134427097>.
>

michaelgregorius commented:
Here's what it looks like without the handles:
![editor-toolbars-nohandle](https://cloud.githubusercontent.com/assets/9293269/9472569/7d44be9c-4b56-11e5-9036-0820dfb13e66.png)
The problem now is that the handles acted as separators and if you take them away it gets a bit crowded in the toolbar area. Adding spacer widgets at the end of the toolbar also feels wrong because logically the space would have to be in between the toolbars.

Shouldn't it be possible to style the handles via style sheets and to keep them?

@grejppi Calm down! :wink: I agree that the handles are not a thing of beauty but this is about trying to improve things by taking one step after another.

Wallacoloo commented:
@michaelgregorius Does the "Lock toolbars" button seem feasible? If so, it gives those of us who prefer the "crowded" look that possibility, while also allowing to keep some form of separation for those who prefer that, based on whether the toolbars are "locked" or not.

michaelgregorius commented:
@Wallacoloo Yes, I guess something like that could be implemented. I just don't know in how far Qt's code will honor all these settings. Let's for example assume that we have one locked toolbar in the middle of the toolbar area and there are other movable toolbars. What will happen if we move these toolbars in the area of the locked one? I also don't know whether Qt allows for non-movable bars to be in the middle of the area.

grejppi commented:
I just don't think this would make sense, especially in the B+B editor. Currently there are two distinct groups of buttons: on the right, the ones that deal with steps; and on the left, the ones that deal with tracks. It provides a visual clue about their relationship, and makes it almost impossible to click the wrong button.

Now in this design they are all grouped together, which requires a lot more effort to pick the right button.

michaelgregorius commented:
@grejppi I agree that the grouping was more distinct in the unchanged version. However, this might also be an indicator that the icons cannot be distinguished very well (most of them are blueish grey).

To provide an even more user friendly experience it might even be better to get rid of the tool buttons and to provide these actions in the areas where they are used, e.g. by having an add track button beneath the tracks similar to the button in the channel mixer. But that would be a lot more effort.

grejppi commented:
Why are you bringing icons to this? A different icon set wouldn't change the fact that the buttons are so close together your mouse will have to make smaller movements to get to the desired one.

In fact, this is the reason why the add buttons are on the right and the remove button is on the left.

I was the one who moved the buttons there [in the first place](https://github.com/LMMS/lmms/pull/56). Initially I had placed the buttons in the opposite order (add on the left, remove on the right), but every time I moved the cursor to the top right corner to add steps I had to move the cursor back a tiny amount because I missed the button and was about to hit remove instead.

Then I swapped them and it was perfect.

It makes sense to have the buttons on the right for a simple reason: the patterns extend to the right. And usually one starts filling the pattern from the beginning (left) and moving to the right. Run out of steps? No problem, the add button is right in the direction where you're going. If this pull request is merged, you'll have to move your cursor back just to click one button, interrupting your workflow.

michaelgregorius commented:
@grejppi I brought the icons into this because I interpreted "a lot more effort to pick the right button" as "the buttons cannot be differentiated if they are so close to each other" and this might be caused by icons that are all very similar looking. I might have interpreted that wrong.

I agree with you that the separation made sense from a user's point of view: tracks are shown on the left so the buttons are on the left. Steps are on the right so their buttons are shown on the right. I think I have an idea how to solve this: I will make the two separate toolbars for track actions and step actions one toolbar again and reintroduce the stretch in that toolbar.

michaelgregorius commented:
Reintroduced the stretch. Current state:
![editor-toolbars-correctedbbeditor](https://cloud.githubusercontent.com/assets/9293269/9476928/b53f1950-4b6f-11e5-86f1-e97cafd6ec13.png)

tresf commented:
> I just don't think this would make sense, especially in the B+B editor. Currently there are two distinct groups of buttons: on the right, the ones that deal with steps; and on the left, the ones that deal with tracks. It provides a visual clue about their relationship, and makes it almost impossible to click the wrong button.
>
> Now in this design they are all grouped together, which requires a lot more effort to pick the right button.

Couldn't this all haven summarized "Can we still float `Add /remove BB steps` to the right?"

I feel like there's some form of spiteful territorial undertone to this PR and it gets us nowhere.

I had originally liked this proposed idea because I thought it fixed the stacking/overflow problem illustrated in #387, but now that I realize it doesn't do overflowing at all, the only benefit is the ability to hide unused components and I think that needs to be weighed against the strange undesirable collapsing behavior that it introduces via https://github.com/LMMS/lmms/pull/2287#issuecomment-134427097, 2nd screenshot.

-Tres

michaelgregorius commented:
@tresf It also fixes #2286 by adding the second row of toolbars. :wink:

To get rid of the collapsing behavior and to implement the overflow the current toolbar implementation would likely have to be replaced with something custom.

StakeoutPunch commented:
Just a quick question: are those screenshots or markups? The gradient is not constant throughout the toolbar, resetting for each cluster of buttons. It really breaks the look.

michaelgregorius commented:
@StakeoutPunch I guess that's somehow caused by the style sheet. I will have to check whether it behaves more consistent with the stylesheet disabled. If it also looks strange in that case then it is likely a Qt issue or caused by some code of the LMMS widgets.

michaelgregorius commented:
I have fixed the gradient problem and have also fixed the problem with the timeline controls which had a smaller size than all other buttons.

Furthermore, I took the liberty to adjust the CSS for the `QToolBar` and the `QToolButton`. This is the current state of the pull request:
![editor-toolbars-css-new](https://cloud.githubusercontent.com/assets/9293269/9701096/8158210e-541b-11e5-9577-69b98bc7d19b.png)
* Before these adjustments all tool buttons looked like they were pressed which made it for example hard to distinguish which mode is selected in the piano roll editor.
* The hover effect is now a bit more subtle so that it looks not as intensive as when a button is finally pressed / checked.
* I have adjusted the radii of the play and stop buttons to be symmetric so that they do not look as "skewed" as before.
* The normal buttons also have a bit more radius now.

Wallacoloo commented:
@michaelgregorius :+1: on your CSS changes.

tresf commented:
Removing fixed sizes :+1: 
Better indication when pressed :+1: 
Overall style :+1:

michaelgregorius commented:
@Wallacoloo, @tresf Shall I rebase, squash and merge then?

tresf commented:
I think you've addressed everyone's wants/needs, made it look better and given plenty of time for feedback.  Good in my book.
